pennylane/compiler/python_compiler/xdsl_extras/constraints.py

```diff
@@ -16,76 +16,147 @@
+from typing import TypeVar
+
+from xdsl.dialects.builtin import ArrayAttr, IntAttr, IntAttrConstraint, MemRefType, TensorType
+from xdsl.ir import Attribute, AttributeInvT
+from xdsl.irdl import (
+    AnyAttr,
+    AnyInt,
+    AttrConstraint,
+    ConstraintContext,
+    EqIntConstraint,
+    IntSetConstraint,
+    IRDLAttrConstraint,
+    RangeLengthConstraint,
+    RangeOf,
+)
 from xdsl.utils.exceptions import VerifyException
 
 
-@dataclass(frozen=True)
-class _RankConstraint(AttrConstraint, ABC):
-    """Internal base class to constrain an attribute to be of a given rank.
+@dataclass(frozen=True, init=False)
+class ContainerConstraint(AttrConstraint, ABC):
+    r"""Internal base class for constraining the element type and shape of container types.
+
+    The shape of the container can be constrained by providing a constraint either directly
+    for the shape, or by providing a constraint for the rank. There are two ways to provide
+    an explicit shape constraint:
+
+        * By providing an ``IRDLAttrConstraint`` for the ``shape`` argument.
+        * By providing a sequence of ``int``\ s specifying the concrete expected shape for
+          the ``shape`` argument.
+
+    There are three ways to provide the rank constraint:
+
+        * By providing an ``IRDLAttrConstraint`` for the ``rank`` argument.
+        * By providing an ``int`` representing the concrete expected rank for the ``rank``
+          argument.
+        * By providing a sequence of ``int``\ s specifying the various allowed ranks for the
+          ``rank`` argument.
 
-    Subclasses must provide 'expected_type' and 'type_name'.
+    .. note::
+
+        Either the ``shape`` or ``rank`` constraint must be provided, not both.
+
+    Args:
+        element_type (IRDLAttrConstraint | None): The constraint for the element type.
+            Default is ``None``, which indicates that any element type is allowed.
+        shape (IRDLAttrConstraint | Sequence[int] | None): The constraint for the shape.
+            Default is ``None``, which indicates that any shape is allowed.
+        shape (IRDLAttrConstraint | Sequence[int] | int | None): The constraint for the
+            rank. Default is ``None``, which indicates that any rank is allowed.
     """
 
-    expected_rank: int
-    """The expected rank."""
+    element_type: IRDLAttrConstraint[AttributeInvT]
+
+    shape: IRDLAttrConstraint[AttributeInvT]
+
+    def __init__(
+        self,
+        *,
+        element_type: IRDLAttrConstraint[AttributeInvT] | None = None,
+        shape: IRDLAttrConstraint[AttributeInvT] | Iterable[int] | None = None,
+        rank: IRDLAttrConstraint[AttributeInvT] | Iterable[int] | int | None = None,
+    ):
+        element_type = element_type or AnyAttr()
+        element_type_constr = element_type
+        shape_constr = None
+
+        if shape is not None and rank is not None:
+            raise ValueError("Only one of 'shape' or 'rank' must be provided.")
```

Comment:
```suggestion
        if shape is not None and rank is not None:
            raise ValueError("Only one of 'shape' or 'rank' may be provided.")
```

I think this is more correct, since we also have the option to provide neither `shape` nor `rank`.